src/wallet/wallet_tx_builder.cpp

```diff
@@ -303,25 +508,38 @@ InputSelectionResult WalletTxBuilder::ResolveInputsAndPayments(
     for (const auto& payment : payments) {
         sendAmount += payment.GetAmount();
     }
-    CAmount targetAmount{sendAmount - fee};
 
-    // TODO: the set of recipient pools is not quite sufficient information here; we should
-    // probably perform note selection at the same time as we're performing resolved payment
-    // construction above.
-    if (!spendableMut.LimitToAmount(targetAmount, dustThreshold, resolved.GetRecipientPools())) {
+    CAmount finalFee;
+    CAmount targetAmount;
+    if (fee.has_value()) {
+        finalFee = fee.value();
+        targetAmount = sendAmount + finalFee;
+        // TODO: the set of recipient pools is not quite sufficient information here; we should
+        // probably perform note selection at the same time as we're performing resolved payment
+        // construction above.
+        bool foundSufficientFunds = spendableMut.LimitToAmount(
+                sendAmount + finalFee,
+                dustThreshold,
+                resolved.GetRecipientPools());
         CAmount changeAmount{spendableMut.Total() - targetAmount};
-        return InvalidFundsError(
-                spendableMut.Total(),
-                changeAmount > 0 && changeAmount < dustThreshold
-                // TODO: we should provide the option for the caller to explicitly
-                // forego change (definitionally an amount below the dust amount)
-                // and send the extra to the recipient or the miner fee to avoid
-                // creating dust change, rather than prohibit them from sending
-                // entirely in this circumstance.
-                // (Daira disagrees, as this could leak information to the recipient
-                // or publicly in the fee.)
-                ? InvalidFundsReason(DustThresholdError(dustThreshold, changeAmount))
-                : InvalidFundsReason(InsufficientFundsError(targetAmount)));
+        if (!foundSufficientFunds) {
+            return ReportInvalidFunds(
+                    spendableMut,
+                    false,
+                    finalFee,
+                    dustThreshold,
+                    targetAmount,
+                    changeAmount);
+        }
+    } else {
+        auto limit_result = IterateLimit(sendAmount, dustThreshold, spendableMut, resolved, ChangeAddressPool(selector));
+        if (limit_result.has_value()) {
+            spendableMut = limit_result.value().first;
+            finalFee = limit_result.value().second;
```

Comment:
```suggestion
            std::tie(spendableMut, finalFee) = limit_result.value();
```

Reply:
This is also fixed in the final commit of this PR.